e2e/internal/upgrade/upgrade_test.go

```diff
@@ -0,0 +1,353 @@
+//go:build e2e
+
+/*
+Copyright (c) Edgeless Systems GmbH
+
+SPDX-License-Identifier: AGPL-3.0-only
+*/
+
+package upgrade
+
+import (
+	"context"
+	"errors"
+	"flag"
+	"fmt"
+	"os"
+	"os/exec"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/edgelesssys/constellation/v2/e2e/internal/kubectl"
+	"github.com/edgelesssys/constellation/v2/internal/config"
+	"github.com/edgelesssys/constellation/v2/internal/constants"
+	"github.com/edgelesssys/constellation/v2/internal/file"
+	"github.com/edgelesssys/constellation/v2/internal/semver"
+	"github.com/spf13/afero"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	coreV1 "k8s.io/api/core/v1"
+	metaV1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+var (
+	targetImage         = flag.String("target-image", "", "Image (shortversion) to upgrade to.")
+	targetKubernetes    = flag.String("target-kubernetes", "", "Kubernetes version (MAJOR.MINOR.PATCH) to upgrade to. Defaults to default version of target CLI.")
+	targetMicroservices = flag.String("target-microservices", "", "Microservice version (MAJOR.MINOR.PATCH) to upgrade to. Defaults to default version of target CLI.")
+	// When executing the test as a bazel target the workspace path is supplied through an env variable that bazel sets.
+	workspace = flag.String("workspace", "", "Constellation workspace in which to run the tests.")
+	// When executing the test as a bazel target the CLI path is supplied through an env variable that bazel sets.
+	// When executing via `go test` extra care should be taken that the supplied CLI is built on the same commit as this test.
+	cliPath     = flag.String("cli", "", "Constellation CLI to run the tests.")
+	wantWorker  = flag.Int("want-worker", 0, "Number of wanted worker nodes.")
+	wantControl = flag.Int("want-control", 0, "Number of wanted control nodes.")
+)
+
+// TestMain checks that the prerequisites for the test are met:
+// - a workspace is set
+// - a CLI path is set
+// - the constellation-upgrade folder does not exist
+func TestMain(m *testing.M) {
+	workingDir, err := workingDir()
+	if err != nil {
+		fmt.Println(err)
+		os.Exit(1)
+	}
+
+	fmt.Printf("settings workspace to: %s\n", workingDir)
+	if err := os.Chdir(workingDir); err != nil {
+		os.Exit(1)
+	}
+
+	if _, err := getCLIPath(); err != nil {
+		fmt.Println(err)
+		os.Exit(1)
+	}
+	if _, err := os.Stat("constellation-upgrade"); err == nil {
+		fmt.Println("Please remove the existing constellation-upgrade folder.")
+		os.Exit(1)
+	}
+
+	os.Exit(m.Run())
+}
+
+// TestUpgrade checks that the workspace's kubeconfig points to a healthy cluster,
+// we can write an upgrade config, we can trigger an upgrade
+// and the cluster eventually upgrades to the target version.
+func TestUpgrade(t *testing.T) {
+	assert := assert.New(t)
+	require := require.New(t)
+	k, err := kubectl.New()
+	require.NoError(err)
+	assert.NotNil(k)
```

Comment:
```suggestion
	require.NotNil(k)
```
We can't continue without kubectl, right?